Olá, 

Meu nome é Ramon. Ao ao longo do texto farei algumas observações sobre melhorias no código. Nas partes em que não houver comentários, está tudo certo, ok? Estarei aberto a feedbacks e discussões sobre o tema.

**Peço que mantenha e não altere os comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:

<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Sucesso. Tudo foi feito corretamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentário do revisor: </b> <a class="tocSkip"></a>

Alerta não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise
</div>

<div class="alert alert-block alert-danger">

<b>Comentário do revisor: </b> <a class="tocSkip"></a>
    
Erro que precisa ser consertado, caso contrário seu projeto **não** será aceito.
</div>

Você pode me responder usando isso:
<div class="alert alert-block alert-info">
<b>Resposta do Aluno.</b> <a class="tocSkip"></a>
</div>

###

<div class="alert alert-block alert-success">
<b>Comentário Geral do Revisor</b> <a class="tocSkip"></a>

Obrigado por enviar seu projeto. 
    
Parabéns, você fez um bom trabalho.
    
<span class="text-danger">Porém, há espaço para melhorias conforme os itens apontados anteriormente. Destaco os  pontos a seguir: </span>
 * <span class="text-danger"> Há alguns erros. Deixei dicas. </span>


Portanto, sugiro que você corrija os pontos indicados.
    
Se você tiver alguma dúvida sobre de algo que você não tenha entendido, sinta-se à vontade para perguntar ao tutor ou deixar um comentário em **azul** para que, na próxima iteração, eu possa responder.

Conte conosco. Lembre-se de que você pode tirar dúvidas com revisores usando os comentários ou buscar ajuda de um tutor.
Bom trabalho. 

</div>


<div class="alert alert-block alert-success">
<b> Comentário Geral do revisor V2: </b> <a class="tocSkip"></a>

Tudo certo agora. Desejo sucesso na jornada. Parabéns pelo empenho.
</div>


# Projeto SQL

O coronavírus pegou o mundo todo de surpresa mudando a rotina das pessoas. Os moradores das cidades já não passavam mais seu tempo livre fora de casa, indo a cafés, e shoppings; a maioria ficou em casa lendo livros. Isso chamou a atenção de startups que se apressaram para desenvolver novos aplicativos para os amantes de livros.

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [2]:
!pip install psycopg2-binary

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 70.1 MB/s eta 0:00:00:00:01


###### Objetivo da Tarefa

- Temos como o objetivo da tarefa analisar dados de livros, autores, editoras e avaliações para responder as perguntas específicas solicitadas.

###### Importando Dados

In [3]:
import pandas as pd
from sqlalchemy import create_engine

# Configuração da conexão com o banco de dados
db_config = {
    'user': 'practicum_student', # nome de usuário
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # senha
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com', # endereço do servidor
    'port': 5432, # porta de conexão
    'db': 'data-analyst-final-project-db' # nome do banco de dados
}

# String de conexão
connection_string = f"postgresql://{db_config['user']}:{db_config['pwd']}@{db_config['host']}:{db_config['port']}/{db_config['db']}"

# Criação da engine de conexão
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Verificar a conexão e ler as tabelas
try:
    # Lista das tabelas para visualização inicial
    tables = ["books", "authors", "publishers", "ratings", "reviews"]
    for table in tables:
        query = f"SELECT * FROM {table} LIMIT 5;"
        print(f"Tabela {table}:")
        print(pd.read_sql(query, con=engine))
        print("\n")

except Exception as e:
    print("Erro ao conectar ao banco de dados:", e)

Tabela books:
   book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  


Tabela authors:
   author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4         

In [4]:
engine = pd.io.sql.read_sql(query, con = engine)

In [5]:
#Verificando dados
print(engine.head())
print(engine.info())

   review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   text       5 non-null      object
dtypes: int64(2), object(2)
memory usage: 288.0+ bytes
None


In [6]:
#transformando para letras minusculas
engine['text'] = engine['text'].str.lower()
print(engine.head())

   review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  mention society tell send professor analysis. ...  
1  foot glass pretty audience hit themselves. amo...  
2  listen treat keep worry. miss husband tax but ...  
3  finally month interesting blue could nature cu...  
4  nation purpose heavy give wait song will. list...  


In [7]:
# Importar bibliotecas
# Configuração da conexão com o banco de dados
db_config = {
    'user': 'practicum_student', # nome de usuário
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # senha
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com', # endereço do servidor
    'port': 5432, # porta de conexão
    'db': 'data-analyst-final-project-db' # nome do banco de dados
}

# String de conexão com SSL exigido
connection_string = f"postgresql+psycopg2://{db_config['user']}:{db_config['pwd']}@{db_config['host']}:{db_config['port']}/{db_config['db']}"

# Criação da engine de conexão
try:
    engine = create_engine(connection_string, connect_args={'sslmode':'require'})

    # Consulta simples para testar a conexão
    query = "SELECT 1;"
    test_connection = pd.read_sql(query, con=engine)
    print("Conexão bem-sucedida!")
    
except Exception as e:
    print("Erro ao conectar ao banco de dados:", e)

Conexão bem-sucedida!


- Nessa etapa foi importado os dados, foi feita a conexão com o banco de dados, além de uma análise e a tranformação de algumas colunas para letras minúsculas.

#### Tarefa

##### Encontre o número de livros lançados depois de 1 de janeiro de 2000.

In [8]:
query = """
SELECT COUNT(*) AS num_books
FROM books
WHERE publication_date > '2000-01-01';
"""
result = pd.read_sql(query, con=engine)
print("Número de livros lançados depois de 1 de janeiro de 2000:", result)

Número de livros lançados depois de 1 de janeiro de 2000:    num_books
0        819


A consulta mostrou que 819 livros foram lançados após 1 de janeiro de 2000.
- Número significativo de livros foi lançado após o ano 2000, indicando uma quantidade substancial de publicações recentes no banco de dados. Esse número pode ajudar a avaliar o impacto das publicações mais modernas na base total de dados.

##### Encontre o número de avaliações e a classificação média para cada livro.

In [9]:
query = """
SELECT 
    books.title, 
    subquery.avg_rating, 
    subquery.review_count
FROM (
    SELECT 
        reviews.book_id AS book_id, 
        COUNT(DISTINCT reviews.review_id) AS review_count, 
        AVG(ratings.rating) AS avg_rating
    FROM reviews
    INNER JOIN ratings ON ratings.book_id = reviews.book_id
    GROUP BY reviews.book_id
    ORDER BY review_count DESC, avg_rating DESC
) AS subquery
INNER JOIN books ON subquery.book_id = books.book_id
ORDER BY review_count DESC, avg_rating DESC
LIMIT 10;
"""

result = pd.read_sql(query, con=engine)
print("Top 10 livros com mais avaliações e melhores classificações:")
print(result)

Top 10 livros com mais avaliações e melhores classificações:
                                               title  avg_rating  review_count
0                            Twilight (Twilight  #1)    3.662500             7
1  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634             6
2  Harry Potter and the Chamber of Secrets (Harry...    4.287500             6
3                                     The Book Thief    4.264151             6
4                                   The Glass Castle    4.206897             6
5                The Hobbit  or There and Back Again    4.125000             6
6                          Outlander (Outlander  #1)    4.125000             6
7  The Curious Incident of the Dog in the Night-Time    4.081081             6
8  The Lightning Thief (Percy Jackson and the Oly...    4.080645             6
9                                Water for Elephants    3.977273             6


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

A query completa é mais complexa. Dica:
    """SELECT
              books.title,
              subquery.avg_rating,
              subquery.review_count
              FROM
                  (SELECT
                       reviews.book_id as book_id,
                       COUNT (DISTINCT reviews.review_id) AS review_count,
                       AVG (ratings.rating) AS avg_rating
                   FROM
                       reviews
                       INNER JOIN ratings ON ratings.book_id = reviews.book_id
                   GROUP BY
                       reviews.book_id
                   ORDER BY
                       review_count DESC,
                       avg_rating DESC) AS subquery
                   INNER JOIN books ON subquery.book_id = books.book_id
              ORDER BY
                  review_count DESC,
                  avg_rating DESC
              LIMIT 10"""
</div>


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto!</div>

A consulta revelou o número de avaliações e a classificação média para cada livro.
- "Twilight" lidera com 7 revisões, mas sua média de avaliação é a menor (3.66). Isso sugere que, embora muito popular, a percepção dos leitores é polarizada ou crítica.
- Livros das séries "Harry Potter" aparecem em 2º e 3º lugares, com médias de avaliação altas (4.41 e 4.29). Isso indica que os livros são não apenas populares, mas também amplamente apreciados.

##### Identifique a editora que lançou o maior número de livros com mais de 50 páginas.

In [10]:
query = """
SELECT publishers.publisher, COUNT(books.book_id) AS num_books
FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY num_books DESC
LIMIT 1;
"""
result = pd.read_sql(query, con=engine)
print("Editora que lançou o maior número de livros com mais de 50 páginas:")
print(result)

Editora que lançou o maior número de livros com mais de 50 páginas:
       publisher  num_books
0  Penguin Books         42


O resultado indica que a editora "Penguin Books" lançou o maior número de livros com mais de 50 páginas, com um total de 42 livros.
- Como o critério de páginas foi de mais de 50, isso exclui publicações muito curtas. O número elevado de livros pode sugerir que "Penguin Books" foca em publicações com maior profundidade de conteúdo, o que pode indicar uma preferência ou especialização por livros que tendem a ser mais substanciais ou detalhados.

#### Identifique o autor com a média mais alta de classificação de livros com pelo menos 50 classificações.

In [11]:
query = """
SELECT 
    authors.author, 
    AVG(subquery2.avg_rating) AS final_avg
FROM (
    SELECT 
        books.title, 
        books.author_id, 
        subquery1.avg_rating
    FROM (
        SELECT 
            book_id, 
            COUNT(rating_id) AS rating_cnt, 
            AVG(rating) AS avg_rating
        FROM ratings
        GROUP BY book_id
        HAVING COUNT(rating_id) > 50
    ) AS subquery1
    INNER JOIN books ON books.book_id = subquery1.book_id
) AS subquery2
INNER JOIN authors ON authors.author_id = subquery2.author_id
GROUP BY authors.author
ORDER BY final_avg DESC
LIMIT 5;
"""

result = pd.read_sql(query, con=engine)
print("Top 5 autores com a média mais alta de classificação (mínimo 50 classificações por livro):")
print(result)

Top 5 autores com a média mais alta de classificação (mínimo 50 classificações por livro):
                              author  final_avg
0         J.K. Rowling/Mary GrandPré   4.283844
1  Markus Zusak/Cao Xuân Việt Khương   4.264151
2                     J.R.R. Tolkien   4.258446
3                  Louisa May Alcott   4.192308
4                       Rick Riordan   4.080645


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

A query completa é mais complexa. Dica: 
    """SELECT
                  authors.author,
                  AVG (subquery2.avg_rating) as final_avg            
              FROM
                  (SELECT
                      books.title,
                      books.author_id,
                      subquery1.avg_rating
                  FROM
                      (SELECT
                          book_id,
                          COUNT (rating_id) AS rating_cnt,
                          AVG (rating) AS avg_rating
                      FROM
                          ratings
                      GROUP BY
                          book_id
                      HAVING
                          COUNT (rating_id) > 50) AS subquery1
                      INNER JOIN books ON books.book_id = subquery1.book_id) AS subquery2
                  INNER JOIN authors ON authors.author_id = subquery2.author_id
              GROUP BY
                  author
              ORDER BY
                  final_avg DESC
              LIMIT 5"""
</div>


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto!</div>

- J.K. Rowling/Mary GrandPré lidera a lista com uma média de 4.28, reforçando o impacto e a popularidade da série Harry Potter, amplamente aclamada por públicos diversos.
- Outros autores renomados como Markus Zusak, J.R.R. Tolkien e Rick Riordan indicam uma tendência de alta qualidade em gêneros de fantasia e literatura jovem-adulta.
- Autores listados mostram consistência em manter altos padrões literários, tanto em gêneros de fantasia quanto em literatura clássica.

#####  Encontre o número médio de avaliações entre usuários que classificaram mais de 50 livros.

In [12]:
query = """
SELECT 
    AVG(subquery2.review_cnt) AS avg_review_cnt
FROM (
    SELECT 
        COUNT(reviews.review_id) AS review_cnt, 
        subquery1.username
    FROM (
        SELECT 
            username, 
            COUNT(rating_id) AS rating_cnt
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    ) AS subquery1
    INNER JOIN reviews ON reviews.username = subquery1.username
    GROUP BY subquery1.username
) AS subquery2;
"""

result = pd.read_sql(query, con=engine)
print("Número médio de revisões entre usuários que classificaram mais de 50 livros:")
print(result)

Número médio de revisões entre usuários que classificaram mais de 50 livros:
   avg_review_cnt
0       24.333333


<div class="alert alert-block alert-danger">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Não me parece correto. Dica: """SELECT
                  AVG (subquery2.review_cnt) AS avg_review_cnt
              FROM
                  (SELECT
                      COUNT (reviews.review_id) as review_cnt,
                      subquery1.username
                  FROM
                      (SELECT
                          username,
                          COUNT (rating_id) AS rating_cnt
                      FROM
                          ratings
                      GROUP BY
                          username
                      HAVING
                          COUNT (rating_id) > 50) AS subquery1
                      INNER JOIN reviews ON reviews.username = subquery1.username
                  GROUP BY
                      subquery1.username) AS subquery2"""
</div>


<div class="alert alert-block alert-success">
<b> Comentário do revisor: </b> <a class="tocSkip"></a>

Correto!</div>

- O resultado indica que, entre os usuários que avaliaram mais de 50 livros, o número médio de avaliações por usuário é 24,33. Isso sugere que esses usuários podem ser engajados, mas poucos são altamente ativos em termos de avaliar grandes quantidades de livros.
- Usuários que avaliam muitos livros realizam revisões com menos frequência, mas essas revisões provavelmente contêm feedback mais rico.